### PR TITLE
PTY Trait

### PR DESCRIPTION
**Pull request by fabianfreyer**
## Work in progress / please do not merge

This PR builds on #49.

## High level description

This PR adds a trait that contains asynchronous methods for a PTY. This trait is implemented for:
* `AsyncPtyMaster`
* `AsyncPtyMasterReadHalf`
* `AsyncPtyMasterWriteHalf`

## Methods
- [x] `ptsname() -> Poll<OsString,io::Error>`
- [x] `resize(cols: c_ushort, rows: c_ushort) -> Poll<(),io::Error>`
- [x] `winsize() -> Poll<(c_ushort, c_ushort),io::Error>`

### CODE REVIEW

**Comment by pkgw** [comment]
@fabianfreyer It took me a super-long time to get this PR's dependencies in place, but I believe that they've all been merged now. Is this PR ready to go, despite being labeled WIP? If not, are you still interested in pushing it across the finish line?

Cheers!

**Comment by fabianfreyer** [comment]
Hmm, I can't remember what else I wanted to do here, and been using this over at https://github.com/fubarnetes/webterm, so yes, this should be ready to go. I'll open a PR with any further changes later if I remember :)

Thanks!

**Comment by fabianfreyer** [comment]
OTOH, let me still reword that last commit ;)

**Comment by fabianfreyer** [comment]
Now it's ready to go :+1:

**Comment by pkgw** [comment]
@fabianfreyer Thanks for sticking with it!

Besides a style comment that I'm about to make, the CI [failed](https://travis-ci.com/pkgw/stund/jobs/199047277) on macOS:

```
---- src/lib.rs - PtyMaster::resize (line 455) stdout ----
thread 'src/lib.rs - PtyMaster::resize (line 455)' panicked at 'test executable failed:
thread 'main' panicked at 'Could not resize the PTY: Os { 
  code: 25, kind: Other, 
  message: "Inappropriate ioctl for device" }', src/libcore/result.rs:997:5
```

Hopefully that's an easy fix?

File: tokio-pty-process/src/lib.rs [hunk 1]
```diff
@@ -387,6 +387,122 @@ impl AsAsyncPtyFd for AsyncPtyMaster {
...
+    /// ```
+    fn resize(&self, rows: c_ushort, cols: c_ushort) -> Poll<(), io::Error>;
+}
+
+impl<T: AsAsyncPtyFd> PtyMaster for T {
+    fn ptsname(&self) -> Poll<OsString, io::Error> {
+        let mut buf: [libc::c_char; 512] = [0; 512];
+        let fd = try_ready!(self.as_async_pty_fd());
+
+        #[cfg(not(any(target_os = "macos", target_os = "freebsd")))]
```
Comment (pkgw) [inline tokio-pty-process/src/lib.rs]: Here, can you please use the style that I came up with in `tokio-pty-process` [as shown here](https://github.com/pkgw/stund/blob/master/tokio-pty-process/src/lib.rs#L136-L144)? The key being:

```rust
const APPLY_NONBLOCK_AFTER_OPEN: bool = cfg!(target_os = "freebsd");
```

Turning the check into a named constant makes it clearer what the key difference is, *and* reduces tricky `#[cfg()]` logic, and of course it's all the same after the optimizer runs.

Comment (fabianfreyer) [reply tokio-pty-process/src/lib.rs]: Hmm, this does not seem to build on macos anymore:
```
error[E0425]: cannot find function `ptsname_r` in module `libc`
   --> tokio-pty-process/src/lib.rs:478:31
    |
478 |             if unsafe { libc::ptsname_r(fd, buf.as_mut_ptr(), buf.len()) } != 0 {
    |                               ^^^^^^^^^ did you mean `ptsname`?
```

**Comment by fabianfreyer** [comment]
> the CI failed on macOS:
>
>```
>---- src/lib.rs - PtyMaster::resize (line 455) stdout ----
>thread 'src/lib.rs - PtyMaster::resize (line 455)' panicked at 'test executable failed:
>thread 'main' panicked at 'Could not resize the PTY: Os { 
>  code: 25, kind: Other, 
>  message: "Inappropriate ioctl for device" }', src/libcore/result.rs:997:5
>```
> Hopefully that's an easy fix?

Hmm, I can reproduce this, but have no idea what's going on there. I think the problem may be that there is no process attached to the slave end, since with the following patch, the test passes:
```diff
--- a/tokio-pty-process/src/lib.rs
+++ b/tokio-pty-process/src/lib.rs
@@ -431,7 +431,7 @@ pub trait PtyMaster {
     /// extern crate tokio_pty_process;
     /// extern crate libc;
     ///
-    /// use tokio_pty_process::{AsyncPtyMaster, PtyMaster};
+    /// use tokio_pty_process::{AsyncPtyMaster, PtyMaster, CommandExt};
     /// use tokio::prelude::*;
     /// use std::ffi::OsString;
     /// use libc::c_ushort;
@@ -453,6 +453,10 @@ pub trait PtyMaster {
     /// fn main() {
     ///     let master = AsyncPtyMaster::open().expect("Could not open the PTY");
     ///
+    ///     let mut child = std::process::Command::new("cat")
+    ///         .spawn_pty_async(&master)
+    ///         .expect("Could not spawn child");
+    ///
     ///     Resize {
     ///         pty: master,
     ///         cols: 80,
@@ -460,9 +464,44 @@ pub trait PtyMaster {
     ///     }
     ///     .wait()
     ///     .expect("Could not resize the PTY");
+    ///
+    ///     child.kill().expect("Could not kill child");
     /// }
     /// ```
```

I've added a commit that adds this.

**Comment by fabianfreyer** [comment]
rebased on current master.

File: tokio-pty-process/src/lib.rs (hunk 1 above)
Comment (pkgw) [reply tokio-pty-process/src/lib.rs]: Ah, that's too bad! Oh well.

**Comment by pkgw** [comment]
Merged! Shall I cut a new release?

**Comment by fabianfreyer** [comment]
Thanks! Yeah, a new release would be awesome!

**Comment by fabianfreyer** [comment]
(Not sure, but this may warrant a 0.2 release)

**Comment by pkgw** [comment]
@fabianfreyer OK, tokio-pty-process 0.4.0 has just made it into the world! stund 0.1.6 will be a tiny bump that updates the dep. Cf #79.

**Comment by fabianfreyer** [comment]
@pkgw Thanks a lot!

**Pull request merged**